### ISSUE TITLE
Unresolved overloaded function in dependent comma expression

### ISSUE THREAD

**Issue by Sinacam** [llvm/llvm-project#60333]
```c++
void f(int);
void f(double);

struct weirdo {
    void operator,(void(*)(int));
};

auto x = [](auto t) -> decltype(t, f){};
auto y = [](weirdo t) -> decltype(t, f){};
auto z = [](auto t) -> decltype(t + f){};
```

`x` fails to compile with
```
error: reference to overloaded function could not be resolved; did you mean to call it?
```

`t, f` is a dependent expression and well-formed with some `t`, as demonstrated with `y`, so `x` should be well-formed.
Considering that `z` compiles, as well as any other operator, I guess it's a problem with the comma operator.

**Comment by shafik** [comment]
I think clang may be correct here, if a bit aggressive in diagnosing it. See [over.over p5](https://timsong-cpp.github.io/cppwp/over.over#5) which says:

> ... After such eliminations, if any, there shall remain exactly one selected function

if we try and call `x(...)` there can not be a context that will allow us to choose an overload: https://godbolt.org/z/z1qTnM6W3

**Comment by shafik** [comment]
Specifically I believe this is ill-formed no diagnostic required as per [temp.res.general p6](https://timsong-cpp.github.io/cppwp/temp.res.general#6.1) which says:

> The program is ill-formed, no diagnostic required, if
> - no valid specialization can be generated for a template or a substatement of a [constexpr if](https://timsong-cpp.github.io/cppwp/stmt.if#def:constexpr_if) statement within a template and the template is not instantiated, or

**Comment by Sinacam** [comment]
> I think clang may be correct here, if a bit aggressive in diagnosing it. See [over.over p5](https://timsong-cpp.github.io/cppwp/over.over#5) which says:
> 
> > ... After such eliminations, if any, there shall remain exactly one selected function
> 
> if we try and call `x(...)` there can not be a context that will allow us to choose an overload: https://godbolt.org/z/z1qTnM6W3

I think I'm missing something here. The point I'm making with `y` and `weirdo` is precisely that we _can_ choose an overrload: https://godbolt.org/z/4v8c8KE9v

**Comment by shafik** [comment]
> > if we try and call `x(...)` there can not be a context that will allow us to choose an overload: https://godbolt.org/z/z1qTnM6W3
> 
> I think I'm missing something here. The point I'm making with `y` and `weirdo` is precisely that we _can_ choose an overload for `x`: https://godbolt.org/z/4v8c8KE9v

Yes, b/c it has the context that the operand to `void operator,(void(*)(int))` has to take an *int* parameter which allows it to eliminate the *double* choice.

**Comment by Sinacam** [comment]
> Yes, b/c it has the context that the operand to `void operator,(void(*)(int))` has to take an _int_ parameter which allows it to eliminate the _double_ choice.

That's the point, that's a valid specialization for `x`. `x.operator()` specialized on `weirdo` is valid. It's not _"no valid specialization can be generated"_.
